mita/openstack.py

```diff
@@ -157,47 +157,73 @@ def _get_flavor(self) -> NodeSize:
+
+        Args:
+            name:   the name of the network.
+
+        Returns:
+            OpenStackNetwork instance
+
+        Raises:
+            ResourceNotFound when no matching resource was found.
+        """
+        url = f"{self.driver_v2._networks_url_prefix}?name={name}"
+        object_ = self.driver_v2.network_connection.request(url).object
+        networks = self.driver_v2._to_networks(object_)
+
+        if not networks:
+            raise ResourceNotFound(f"No network resource with name {name} found.")
+
+        return networks[0]
+
+    def _has_free_ip_address(self, net: OpenStackNetwork) -> bool:
+        """
+        Return True if the given network has an ip address to lease.
 
         Arguments:
-            network: String, The network UUID
+            net:    The reference to OpenStackNetwork
 
         Returns:
             bool - True on success else False
 
         Note: This method returns True only if the given network has more than 3
         """
         try:
-            resp = self.driver_v2.network_connection.request(
-                "/v2.0/network-ip-availabilities/{}".format(network),
+            url = f"/v2.0/network-ip-availabilities/{net.id}"
+            resp = self.driver_v2.network_connection.request(url)
+            subnets = resp.object.get("network_ip_availability", {}).get(
+                "subnet_ip_availability", []
             )
-        except BaseException as be:
-            logger.info(be)
-            return False
 
-        if resp.status != 200:
-            logger.debug("Failed to gather the network details.")
-            return False
+            for subnet in subnets:
+                _free_ips = subnet.get("total_ips") - subnet.get("used_ips")
 
-        _total = resp.object.get("network_ip_availability", {}).get("total_ips")
-        _used = resp.object.get("network_ip_availability", {}).get("used_ips")
+                if _free_ips > 3:
+                    self.subnet = subnet.get("cidr")
+                    return True
+        except BaseException as be:  # noqa
+            logger.warning(be)
 
-        return (_total - _used) > 3
+        logger.info("%s has no IP addresses to lease.", net.name)
+        return False
 
     def _get_network(self, network: Optional[str] = None) -> OpenStackNetwork:
         """
-        Return the provider network.
+        Return the instance of OpenStackNetwork from the known list or given name.
+
+        A network object is returned only if it contains a free IP address to lease.
 
         Arguments:
-            network: Network to be attached to the VM
+            network: The name of the network to be retrieved.
 
         Returns:
-            An OpenStackNetwork libcloud object
+            An OpenStackNetwork libcloud object.
 
         Raises:
-            GetIPError when there are no suitable networks
+            ResourceNotFound is raised when there are no suitable networks.
         """
         _networks = (
```

Comment:
This is not specifically related to your PR, and we can fix this later in order to keep the scope of this change manageable, but I'll just mention this here: I wonder why this variable is named `_networks` instead of simply `networks`, or more clearly, `network_names`. We could probably rename this (separate PR).

Reply:
Done